examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,1673 @@
+
+    def __call__(self, model_output, timesteps, sample):
+        alphas = extract_into_tensor(self.alphas, timesteps, sample.shape)
+        sigmas = extract_into_tensor(self.sigmas, timesteps, sample.shape)
+        if self.prediction_type == "epsilon":
+            pred_x_0 = (sample - sigmas * model_output) / alphas
+        elif self.prediction_type == "sample":
+            pred_x_0 = model_output
+        elif self.prediction_type == "v_prediction":
+            pred_x_0 = alphas * sample - sigmas * model_output
+        else:
+            raise ValueError(
+                f"Prediction type {self.prediction_type} is not supported; currently, `epsilon`, `sample`, and"
+                f" `v_prediction` are supported."
+            )
+
+        return pred_x_0
+
+
+# Based on SpectralConv1d from the official StyleGAN-T code
+# https://github.com/autonomousvision/stylegan-t/blob/36ab80ce76237fefe03e65e9b3161c040ae888e3/networks/discriminator.py#L29
+class SpectralConv1d(torch.nn.Conv1d):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        torch.nn.utils.SpectralNorm.apply(self, name="weight", n_power_iterations=1, dim=0, eps=1e-12)
+
+
+# Based on ResidualBlock from the official StyleGAN-T code
+# https://github.com/autonomousvision/stylegan-t/blob/36ab80ce76237fefe03e65e9b3161c040ae888e3/networks/shared.py#L20
+class ResidualBlock(torch.nn.Module):
+    def __init__(self, fn: Callable):
+        super().__init__()
+        self.fn = fn
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return (self.fn(x) + x) / np.sqrt(2)
+
+
+# Based on make_block from the official StyleGAN-T code
+# https://github.com/autonomousvision/stylegan-t/blob/36ab80ce76237fefe03e65e9b3161c040ae888e3/networks/discriminator.py#L64
+class DiscHeadBlock(torch.nn.Module):
+    """
+    StyleGAN-T block: SpectralConv1d => BatchNormLocal => LeakyReLU
+    """
+
+    def __init__(
+        self,
+        channels: int,
+        kernel_size: int,
+        num_groups: int = 8,
+        leaky_relu_neg_slope: float = 0.2,
+    ):
+        super().__init__()
+        self.channels = channels
+
+        self.conv = SpectralConv1d(
+            channels,
+            channels,
+            kernel_size=kernel_size,
+            padding=kernel_size // 2,
+            padding_mode="circular",
+        )
+        self.batch_norm = torch.nn.GroupNorm(num_groups, channels)
+        self.act_fn = torch.nn.LeakyReLU(leaky_relu_neg_slope, inplace=True)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.conv(x)
+        x = self.batch_norm(x)
+        x = self.act_fn(x)
+        return x
+
+
+# Based on DiscHead in the official StyleGAN-T implementation
+# https://github.com/autonomousvision/stylegan-t/blob/36ab80ce76237fefe03e65e9b3161c040ae888e3/networks/discriminator.py#L78
+# TODO: implement image conditioning (see Section 3.2 of paper)
+class DiscriminatorHead(torch.nn.Module):
```

Comment:
I'd be keen on experimenting with simpler Discrimantors: 

* Use a plain ViT backbone either from models like CLIP or just regular image classifiers, and then add a discrimination head on top of the backbone. 
* I believe the above should allow for adding additional protection layers. 
* Instead of Hinge loss, we do regular binary cross-entropy with some label-smoothing. 

This should allow for more simplicity, while label-smoothing should provide a good amount of regularization to the mix.

Reply:
As we are implementing the ADD as per the paper here, we should stick with what's used in the paper to re-produce results as closely as possible. This will keep the script simple and in accordance with the paper. A more experimental script can be added in `research_projects`. 

And here using `dino` makes much more sense. It's also a plain vit but has good feature representation than image classifier models. We are already feeding `clip` embeds as conditioning, so using an image encoder from CLIP won't make sense.

Reply:
This was a suggestion for running experiments :D 

Never meant we should do that here.